File: Algorithms_medium/22_Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def dfs(left, right, t):
            if left == n and right == n:
                ans.append(t)
                return
            if left < n:
                dfs(left+1, right, t+'(')
            if right < left:
                dfs(left, right+1, t+')')
        dfs(0,0,'')
        return ans
'''
深度优先搜索算法dfs，先添加左括号，然后扩展右括号，自动循环
'''

File: Algorithms_C++/22_Generate_Parentheses.cpp
class Solution {
    void backtrack(vector<string>& ans, string& cur, int left, int right, int n){
        if (cur.size() == n * 2){
            ans.push_back(cur);
            return;
        }
        if (left < n){
            cur.push_back('(');
            backtrack(ans, cur, left+1, right, n);
            cur.pop_back();
        }
        if (right < left){
            cur.push_back(')');
            backtrack(ans, cur, left, right+1, n);
            cur.pop_back();
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        string current;
        backtrack(result, current, 0, 0, n);
        return result;
    }
};